File: src/memex/api/routers/metrics.py
"""Métricas de costo LLM sobre la tabla `llm_calls`.

Dos endpoints para la vista /metricas:
- `GET /metrics/llm/rollup` — agregaciones server-side (GROUP BY): KPIs, costo por fuente, por
  módulo (de `purpose`), por modelo, matriz fuente x módulo y serie diaria. A diferencia de
  `/finance/expenses` (filas crudas + agregación en cliente), `llm_calls` es denso y append-only,
  así que agregar en SQL evita traer todo al browser.
- `GET /metrics/llm/calls` — auditoría: filas crudas con filtros incluir/excluir multi-valor, orden
  por columna (whitelist) y paginación offset (necesita total para el pager "X-Y de N").

El módulo se deriva de `purpose` con un único CASE (`_MODULE_CASE`) reusado en ambos endpoints. El
bucket diario se fija a una TZ (`_BUCKET_TZ`) para que "hoy" no se parta a medianoche (created_at es
TIMESTAMPTZ). `untabulated` (modelo sin precio) se deriva de datos: tokens>0 con cost_usd=0.
"""

from datetime import UTC, datetime
from typing import Annotated, Any, Literal

from fastapi import APIRouter, Depends, Query
from sqlalchemy import text

from memex.api.auth import current_user_id
from memex.api.schemas import LlmCallList, LlmRollup
from memex.db import connection
from memex.logging import get_logger

router = APIRouter(prefix="/metrics", tags=["metrics"])

UserID = Annotated[int, Depends(current_user_id)]

_log = get_logger("memex.api.metrics")

#: TZ del bucket diario (locale es-MX del dashboard) — alinea "hoy" con el reloj de pared.
_BUCKET_TZ = "America/Mexico_City"

#: Deriva el módulo/etapa desde `purpose`. El literal `extract_grouped` va ANTES del wildcard
#: `extract_%`; un purpose futuro desconocido cae al ELSE y se ve nombrado (no se pierde el gasto).
_MODULE_CASE = """
    CASE
        WHEN purpose = 'module_route' THEN 'routing'
        WHEN purpose LIKE 'summarize%' THEN 'summarize'
        WHEN purpose = 'extract_grouped' THEN 'grouped'
        WHEN purpose LIKE 'extract_%' THEN substring(purpose FROM 'extract_(.*)')
        WHEN purpose LIKE 'calendar%' THEN 'calendar'
        WHEN purpose = 'ocr' THEN 'ocr'
        ELSE purpose
    END
""".strip()

#: Etiqueta de fuente: nombre real, o pseudo-fuente para las llamadas sin source (calendar cruza
#: fuentes; el resto sin atribución = "(sin source)") — así el gasto sin source se VE, no se pierde.
_SOURCE_LABEL = (
    "COALESCE(s.name, CASE WHEN lc.purpose LIKE 'calendar%' "
    "THEN '(calendar)' ELSE '(sin source)' END)"
)

#: Columnas permitidas para ordenar la auditoría (nunca interpolar input del usuario como columna).
_SORT_COLS = {"created_at": "created_at", "cost_usd": "cost_usd", "latency_ms": "latency_ms"}


def _window(since: datetime | None, until: datetime | None, params: dict[str, Any]) -> str:
    """WHERE de ventana sobre `llm_calls lc` (alias lc). Agrega :since/:until a `params`."""
    clauses = ["lc.user_id = :uid"]
    if since is not None:
        clauses.append("lc.created_at >= :since")
        params["since"] = since
    if until is not None:
        clauses.append("lc.created_at < :until")
        params["until"] = until
    return " AND ".join(clauses)


@router.get("/llm/rollup", response_model=LlmRollup)
async def llm_rollup(
    user_id: UserID,
    since: datetime | None = None,
    until: datetime | None = None,
) -> dict[str, Any]:
    """Agregaciones de costo LLM del rango [`since`, `until`) para la vista /metricas."""
    params: dict[str, Any] = {"uid": user_id, "tz": _BUCKET_TZ}
    where = _window(since, until, params)

    with connection() as conn:
        krow = (
            conn.execute(
                text(f"""
                SELECT
                    COUNT(*) AS calls,
                    COALESCE(SUM(lc.cost_usd), 0) AS cost_usd,
                    COALESCE(SUM(lc.prompt_tokens), 0) AS prompt_tokens,
                    COALESCE(SUM(lc.completion_tokens), 0) AS completion_tokens,
                    COALESCE(SUM(lc.cache_hit_tokens), 0) AS cache_hit_tokens,
                    COUNT(*) FILTER (WHERE lc.status = 'error') AS errors,
                    COALESCE(AVG(lc.latency_ms), 0) AS avg_latency_ms
                FROM llm_calls lc
                WHERE {where}
            """),
                params,
            )
            .mappings()
            .one()
        )

        # Periodo anterior de igual longitud, solo si hay `since` (para la variación %).
        prev_cost: float | None = None
        if since is not None:
            until_eff = until if until is not None else datetime.now(UTC)
            span = until_eff - since
            prev_cost = float(
                conn.execute(
                    text(
                        "SELECT COALESCE(SUM(cost_usd), 0) FROM llm_calls "
                        "WHERE user_id = :uid AND created_at >= :ps AND created_at < :pe"
                    ),
                    {"uid": user_id, "ps": since - span, "pe": since},
                ).scalar_one()
            )

        by_source = (
            conn.execute(
                text(f"""
                SELECT lc.source_id,
                       {_SOURCE_LABEL} AS source_name,
                       COUNT(*) AS calls,
                       COALESCE(SUM(lc.prompt_tokens + lc.completion_tokens), 0) AS tokens,
                       COALESCE(SUM(lc.cost_usd), 0) AS cost_usd
                FROM llm_calls lc
                LEFT JOIN sources s ON s.id = lc.source_id
                WHERE {where}
                GROUP BY lc.source_id, source_name
                ORDER BY cost_usd DESC
            """),
                params,
            )
            .mappings()
            .all()
        )

        by_module = (
            conn.execute(
                text(f"""
                SELECT {_MODULE_CASE} AS module,
                       COUNT(*) AS calls,
                       COALESCE(SUM(lc.prompt_tokens + lc.completion_tokens), 0) AS tokens,
                       COALESCE(SUM(lc.cost_usd), 0) AS cost_usd
                FROM llm_calls lc
                WHERE {where}
                GROUP BY module
                ORDER BY cost_usd DESC
            """),
                params,
            )
            .mappings()
            .all()
        )

        by_model = (
            conn.execute(
                text(f"""
                SELECT lc.model,
                       COUNT(*) AS calls,
                       COALESCE(SUM(lc.prompt_tokens), 0) AS prompt_tokens,
                       COALESCE(SUM(lc.completion_tokens), 0) AS completion_tokens,
                       COALESCE(SUM(lc.cost_usd), 0) AS cost_usd,
                       (COALESCE(SUM(lc.prompt_tokens + lc.completion_tokens), 0) > 0
                        AND COALESCE(SUM(lc.cost_usd), 0) = 0) AS untabulated
                FROM llm_calls lc
                WHERE {where}
                GROUP BY lc.model
                ORDER BY cost_usd DESC
            """),
                params,
            )
            .mappings()
            .all()
        )

        by_source_module = (
            conn.execute(
                text(f"""
                SELECT lc.source_id,
                       {_SOURCE_LABEL} AS source_name,
                       {_MODULE_CASE} AS module,
                       COUNT(*) AS calls,
                       COALESCE(SUM(lc.cost_usd), 0) AS cost_usd
                FROM llm_calls lc
                LEFT JOIN sources s ON s.id = lc.source_id
                WHERE {where}
                GROUP BY lc.source_id, source_name, module
                ORDER BY cost_usd DESC
            """),
                params,
            )
            .mappings()
            .all()
        )

        daily_rows = (
            conn.execute(
                text(f"""
                SELECT (lc.created_at AT TIME ZONE :tz)::date AS day,
                       {_MODULE_CASE} AS module,
                       COALESCE(SUM(lc.cost_usd), 0) AS cost_usd
                FROM llm_calls lc
                WHERE {where}
                GROUP BY day, module
                ORDER BY day
            """),
                params,
            )
            .mappings()
            .all()
        )

    # Pivot de la serie diaria (formato largo → {day, total, by_module}).
    daily_map: dict[str, dict[str, Any]] = {}
    for r in daily_rows:
        day = r["day"].isoformat()
        entry = daily_map.setdefault(day, {"day": day, "total": 0.0, "by_module": {}})
        cost = float(r["cost_usd"])
        entry["by_module"][r["module"]] = cost
        entry["total"] += cost
    daily = [daily_map[d] for d in sorted(daily_map)]

    prompt = int(krow["prompt_tokens"])
    cache_hit = int(krow["cache_hit_tokens"])
    calls = int(krow["calls"])
    cost = float(krow["cost_usd"])
    kpis = {
        "cost_usd": cost,
        "calls": calls,
        "prompt_tokens": prompt,
        "completion_tokens": int(krow["completion_tokens"]),
        "cache_hit_tokens": cache_hit,
        "cache_hit_ratio": (cache_hit / prompt) if prompt else 0.0,
        "avg_cost_usd": (cost / calls) if calls else 0.0,
        "avg_latency_ms": float(krow["avg_latency_ms"]),
        "errors": int(krow["errors"]),
        "prev_cost_usd": prev_cost,
    }

    _log.info("metrics.rollup", user_id=user_id, calls=calls, cost_usd=cost)
    return {
        "kpis": kpis,
        "by_source": [
            {
                "source_id": r["source_id"],
                "source_name": r["source_name"],
                "calls": int(r["calls"]),
                "tokens": int(r["tokens"]),
                "cost_usd": float(r["cost_usd"]),
            }
            for r in by_source
        ],
        "by_module": [
            {
                "module": r["module"],
                "calls": int(r["calls"]),
                "tokens": int(r["tokens"]),
                "cost_usd": float(r["cost_usd"]),
            }
            for r in by_module
        ],
        "by_model": [
            {
                "model": r["model"],
                "calls": int(r["calls"]),
                "prompt_tokens": int(r["prompt_tokens"]),
                "completion_tokens": int(r["completion_tokens"]),
                "cost_usd": float(r["cost_usd"]),
                "untabulated": bool(r["untabulated"]),
            }
            for r in by_model
        ],
        "by_source_module": [
            {
                "source_id": r["source_id"],
                "source_name": r["source_name"],
                "module": r["module"],
                "calls": int(r["calls"]),
                "cost_usd": float(r["cost_usd"]),
            }
            for r in by_source_module
        ],
        "daily": daily,
        "modules": [r["module"] for r in by_module],
    }


def _multi_filter(
    col: str,
    name: str,
    values: list[Any] | None,
    mode: str,
    clauses: list[str],
    params: dict[str, Any],
) -> None:
    """Filtro incluir/excluir multi-valor sobre `col` (= ANY) — vacío/None = sin filtro."""
    if not values:
        return
    params[name] = values
    expr = f"{col} = ANY(:{name})"
    clauses.append(f"NOT ({expr})" if mode == "exclude" else expr)


FilterMode = Literal["include", "exclude"]


@router.get("/llm/calls", response_model=LlmCallList)
async def llm_calls(
    user_id: UserID,
    since: datetime | None = None,
    until: datetime | None = None,
    status: Annotated[list[str] | None, Query()] = None,
    status_mode: FilterMode = "include",
    module: Annotated[list[str] | None, Query()] = None,
    module_mode: FilterMode = "include",
    model: Annotated[list[str] | None, Query()] = None,
    model_mode: FilterMode = "include",
    source: Annotated[list[str] | None, Query(description="source_name (incl. pseudo)")] = None,
    source_mode: FilterMode = "include",
    q: str | None = None,
    sort: Literal["created_at", "cost_usd", "latency_ms"] = "created_at",
    dir: Literal["asc", "desc"] = "desc",
    limit: Annotated[int, Query(ge=1, le=200)] = 50,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> dict[str, Any]:
    """Auditoría de `llm_calls`: filas crudas filtrables/ordenables/paginadas (debug).

    Filtros incluir/excluir multi-valor por estado/módulo/modelo/fuente (aislar o excluir);
    `q` busca substring en inbox/request/model/purpose. `sort`/`dir` salen de una whitelist fija.
    """
    params: dict[str, Any] = {"uid": user_id, "limit": limit, "offset": offset}
    where = _window(since, until, params)

    filters: list[str] = []
    _multi_filter("status", "status", status, status_mode, filters, params)
    _multi_filter("module", "module", module, module_mode, filters, params)
    _multi_filter("model", "model", model, model_mode, filters, params)
    _multi_filter("source_name", "source", source, source_mode, filters, params)
    if q:
        params["q"] = f"%{q}%"
        filters.append(
            "(CAST(inbox_id AS TEXT) ILIKE :q OR request_id ILIKE :q "
            "OR model ILIKE :q OR purpose ILIKE :q)"
        )
    filter_sql = (" WHERE " + " AND ".join(filters)) if filters else ""

    col = _SORT_COLS[sort]
    direction = "ASC" if dir == "asc" else "DESC"

    sql = f"""
        WITH c AS (
            SELECT lc.id, lc.created_at, lc.purpose, {_MODULE_CASE} AS module, lc.model,
                   lc.prompt_tokens, lc.completion_tokens, lc.cache_hit_tokens, lc.cost_usd,
                   lc.latency_ms, lc.status, lc.error_message, lc.inbox_id, lc.source_id,
                   lc.request_id, lc.metadata,
                   {_SOURCE_LABEL} AS source_name
            FROM llm_calls lc
            LEFT JOIN sources s ON s.id = lc.source_id
            WHERE {where}
        )
        SELECT id, created_at, purpose, module, model, prompt_tokens, completion_tokens,
               cache_hit_tokens, cost_usd, latency_ms, status, error_message, inbox_id,
               source_id, source_name, metadata, COUNT(*) OVER() AS total
        FROM c{filter_sql}
        ORDER BY {col} {direction}, id {direction}
        LIMIT :limit OFFSET :offset
    """
    with connection() as conn:
        rows = conn.execute(text(sql), params).mappings().all()

    total = int(rows[0]["total"]) if rows else 0
    items = [
        {
            "id": r["id"],
            "created_at": r["created_at"],
            "purpose": r["purpose"],
            "module": r["module"],
            "model": r["model"],
            "prompt_tokens": int(r["prompt_tokens"]),
            "completion_tokens": int(r["completion_tokens"]),
            "cache_hit_tokens": int(r["cache_hit_tokens"]),
            "cost_usd": float(r["cost_usd"]),
            "latency_ms": int(r["latency_ms"]),
            "status": r["status"],
            "error_message": r["error_message"],
            "inbox_id": r["inbox_id"],
            "source_id": r["source_id"],
            "source_name": r["source_name"],
            "metadata": r["metadata"],
        }
        for r in rows
    ]
    return {"items": items, "total": total}

File: frontend/src/data/metrics.ts
// Superficie de MÉTRICAS de costo LLM contra la API real (no mocks). Como `finance.ts`: funciones
// async + transform snake_case → camelCase. A diferencia de finance, el backend YA agrega (GROUP BY
// en /metrics/llm/rollup) porque llm_calls es denso; acá solo traemos el rollup y las filas crudas
// de auditoría (/metrics/llm/calls, con filtros incluir/excluir y orden server-side).

import { apiGet } from "@/lib/api"
import { rangeMs, type RangeKey } from "@/lib/selectors"
import { startOfDayInTz, todayInTz } from "@/lib/timezone"

// ---- Tipos de dominio (camelCase) ---------------------------------------------------------------

export interface LlmKpis {
  costUsd: number
  calls: number
  promptTokens: number
  completionTokens: number
  cacheHitTokens: number
  cacheHitRatio: number
  avgCostUsd: number
  avgLatencyMs: number
  errors: number
  /** Costo del periodo anterior de igual longitud; null si el rango es "todo". */
  prevCostUsd: number | null
  /** #Llamadas del periodo anterior; null si no hay `since`. 0 distingue "previo vacío" del delta. */
  prevCalls: number | null
}

export interface SourceCost {
  sourceId: number | null
  sourceName: string
  calls: number
  tokens: number
  costUsd: number
}

export interface ModuleCost {
  module: string
  calls: number
  tokens: number
  costUsd: number
}

export interface ModelCost {
  model: string
  calls: number
  promptTokens: number
  completionTokens: number
  costUsd: number
  /** tokens>0 con costo 0 → modelo sin precio tabulado (gasto silencioso a señalar). */
  untabulated: boolean
}

export interface SourceModuleCost {
  sourceId: number | null
  sourceName: string
  module: string
  calls: number
  costUsd: number
}

export interface DailyCost {
  day: string // 'YYYY-MM-DD'
  total: number
  byModule: Record<string, number>
}

export interface LlmRollup {
  kpis: LlmKpis
  bySource: SourceCost[]
  byModule: ModuleCost[]
  byModel: ModelCost[]
  bySourceModule: SourceModuleCost[]
  daily: DailyCost[]
  /** Módulos presentes en el rango (orden por costo) → series estables del área apilada. */
  modules: string[]
}

export interface LlmCallRow {
  id: number
  createdAt: string
  purpose: string
  module: string
  model: string
  promptTokens: number
  completionTokens: number
  cacheHitTokens: number
  costUsd: number
  latencyMs: number
  status: string
  errorMessage: string | null
  inboxId: number | null
  sourceId: number | null
  sourceName: string | null
  /** Metadata de la fase: extracción {items, discarded, n, ...}; ruteo {chosen, ...}. */
  metadata: Record<string, unknown> | null
}

// ---- Ventana temporal ---------------------------------------------------------------------------

/** Ventana [since, until) en ISO; ambos opcionales (omitir = sin límite por ese lado). `tz` (IANA)
 *  es la zona de display: el cliente computa "hoy"/custom como medianoche-en-`tz` y la pasa al
 *  backend para que el bucket diario coincida con el reloj de pared del usuario. */
export interface MetricsWindow {
  since?: string
  until?: string
  tz?: string
}

export type RangePreset = "today" | "7d" | "30d" | "90d" | "all"

const DAY = 86_400_000

/** Ventana de un preset, anclada a `tz`. "today" arranca en la medianoche de `tz` (no la local del
 *  navegador); 7d/30d/90d son instantes relativos (TZ-agnósticos); "all" no acota. */
export function presetWindow(preset: RangePreset, tz: string): MetricsWindow {
  const now = Date.now()
  switch (preset) {
    case "today": {
      const { y, m, d } = todayInTz(tz)
      return { since: startOfDayInTz(y, m, d, tz), tz }
    }
    case "7d":
      return { since: new Date(now - 7 * DAY).toISOString(), tz }
    case "30d":
      return { since: new Date(now - 30 * DAY).toISOString(), tz }
    case "90d":
      return { since: new Date(now - 90 * DAY).toISOString(), tz }
    case "all":
      return { tz }
  }
}

/** Mapea el RangeKey del picker GLOBAL (topbar) a una ventana — lo usa la vista Resumen. */
export function rangeKeyWindow(range: RangeKey): MetricsWindow {
  const span = rangeMs(range)
  if (!Number.isFinite(span)) return {} // "all"
  return { since: new Date(Date.now() - span).toISOString() }
}

/** Ventana de un rango personalizado (inputs date YYYY-MM-DD) anclada a `tz`. Cada día arranca en
 *  la medianoche de `tz` (no la local del navegador); `until` se vuelve exclusivo del día siguiente
 *  para incluir el día completo (el backend filtra created_at < until). */
export function customWindow(sinceDay: string | undefined, untilDay: string | undefined, tz: string): MetricsWindow {
  const w: MetricsWindow = { tz }
  if (sinceDay) {
    const [y, m, d] = sinceDay.split("-").map(Number)
    w.since = startOfDayInTz(y, m, d, tz)
  }
  if (untilDay) {
    const [y, m, d] = untilDay.split("-").map(Number)
    w.until = new Date(new Date(startOfDayInTz(y, m, d, tz)).getTime() + DAY).toISOString()
  }
  return w
}

// ---- API rows (snake_case) + transforms ---------------------------------------------------------

interface KpisApi {
  cost_usd: number
  calls: number
  prompt_tokens: number
  completion_tokens: number
  cache_hit_tokens: number
  cache_hit_ratio: number
  avg_cost_usd: number
  avg_latency_ms: number
  errors: number
  prev_cost_usd: number | null
  prev_calls: number | null
}
interface SourceApi { source_id: number | null; source_name: string; calls: number; tokens: number; cost_usd: number }
interface ModuleApi { module: string; calls: number; tokens: number; cost_usd: number }
interface ModelApi { model: string; calls: number; prompt_tokens: number; completion_tokens: number; cost_usd: number; untabulated: boolean }
interface SourceModuleApi { source_id: number | null; source_name: string; module: string; calls: number; cost_usd: number }
interface DailyApi { day: string; total: number; by_module: Record<string, number> }
interface RollupApi {
  kpis: KpisApi
  by_source: SourceApi[]
  by_module: ModuleApi[]
  by_model: ModelApi[]
  by_source_module: SourceModuleApi[]
  daily: DailyApi[]
  modules: string[]
}
interface CallApi {
  id: number
  created_at: string
  purpose: string
  module: string
  model: string
  prompt_tokens: number
  completion_tokens: number
  cache_hit_tokens: number
  cost_usd: number
  latency_ms: number
  status: string
  error_message: string | null
  inbox_id: number | null
  source_id: number | null
  source_name: string | null
  metadata: Record<string, unknown> | null
}

function toRollup(r: RollupApi): LlmRollup {
  return {
    kpis: {
      costUsd: r.kpis.cost_usd,
      calls: r.kpis.calls,
      promptTokens: r.kpis.prompt_tokens,
      completionTokens: r.kpis.completion_tokens,
      cacheHitTokens: r.kpis.cache_hit_tokens,
      cacheHitRatio: r.kpis.cache_hit_ratio,
      avgCostUsd: r.kpis.avg_cost_usd,
      avgLatencyMs: r.kpis.avg_latency_ms,
      errors: r.kpis.errors,
      prevCostUsd: r.kpis.prev_cost_usd,
      prevCalls: r.kpis.prev_calls,
    },
    bySource: r.by_source.map((s) => ({
      sourceId: s.source_id,
      sourceName: s.source_name,
      calls: s.calls,
      tokens: s.tokens,
      costUsd: s.cost_usd,
    })),
    byModule: r.by_module.map((m) => ({ module: m.module, calls: m.calls, tokens: m.tokens, costUsd: m.cost_usd })),
    byModel: r.by_model.map((m) => ({
      model: m.model,
      calls: m.calls,
      promptTokens: m.prompt_tokens,
      completionTokens: m.completion_tokens,
      costUsd: m.cost_usd,
      untabulated: m.untabulated,
    })),
    bySourceModule: r.by_source_module.map((c) => ({
      sourceId: c.source_id,
      sourceName: c.source_name,
      module: c.module,
      calls: c.calls,
      costUsd: c.cost_usd,
    })),
    daily: r.daily.map((d) => ({ day: d.day, total: d.total, byModule: d.by_module })),
    modules: r.modules,
  }
}

function toRow(r: CallApi): LlmCallRow {
  return {
    id: r.id,
    createdAt: r.created_at,
    purpose: r.purpose,
    module: r.module,
    model: r.model,
    promptTokens: r.prompt_tokens,
    completionTokens: r.completion_tokens,
    cacheHitTokens: r.cache_hit_tokens,
    costUsd: r.cost_usd,
    latencyMs: r.latency_ms,
    status: r.status,
    errorMessage: r.error_message,
    inboxId: r.inbox_id,
    sourceId: r.source_id,
    sourceName: r.source_name,
    metadata: r.metadata,
  }
}

export function windowParams(qs: URLSearchParams, w: MetricsWindow): void {
  if (w.since) qs.set("since", w.since)
  if (w.until) qs.set("until", w.until)
  if (w.tz) qs.set("tz", w.tz)
}

/** Rollup de costo LLM del rango (KPIs + cortes por fuente/módulo/modelo + matriz + serie diaria). */
export async function fetchLlmRollup(w: MetricsWindow = {}): Promise<LlmRollup> {
  const qs = new URLSearchParams()
  windowParams(qs, w)
  return toRollup(await apiGet<RollupApi>(`/metrics/llm/rollup?${qs.toString()}`))
}

// ---- Apify (gasto real de scraping social: tabla apify_runs) ------------------------------------

export interface ApifyKpis {
  costUsd: number
  runs: number
  itemsScraped: number
  itemsKept: number
  /** Runs con status != ok (error + timeout) — pudieron cobrar igual. */
  errors: number
  /** Cuentas seguidas distintas con actividad en el rango. */
  accounts: number
  prevCostUsd: number | null
  prevRuns: number | null
}

export interface ApifySourceCost {
  sourceId: number | null
  sourceName: string
  runs: number
  itemsScraped: number
  costUsd: number
}

export interface ApifyAccountCost {
  platform: string
  account: string
  runs: number
  itemsScraped: number
  costUsd: number
}

export interface ApifyPlatformCost {
  platform: string
  runs: number
  itemsScraped: number
  costUsd: number
}

export interface ApifyDailyCost {
  day: string
  total: number
  byPlatform: Record<string, number>
}

export interface ApifyRollup {
  kpis: ApifyKpis
  bySource: ApifySourceCost[]
  byAccount: ApifyAccountCost[]
  byPlatform: ApifyPlatformCost[]
  daily: ApifyDailyCost[]
  platforms: string[]
}

export interface ApifyRunRow {
  id: number
  createdAt: string
  platform: string
  account: string
  actorId: string
  apifyRunId: string | null
  status: string
  itemsScraped: number
  itemsKept: number
  /** null = Apify aún no asentó el costo del run. */
  costUsd: number | null
  chargedEvents: Record<string, number> | null
  sourceId: number | null
  sourceName: string | null
  ingestionRunId: string | null
}

interface ApifyKpisApi {
  cost_usd: number
  runs: number
  items_scraped: number
  items_kept: number
  errors: number
  accounts: number
  prev_cost_usd: number | null
  prev_runs: number | null
}
interface ApifySourceApi { source_id: number | null; source_name: string; runs: number; items_scraped: number; cost_usd: number }
interface ApifyAccountApi { platform: string; account: string; runs: number; items_scraped: number; cost_usd: number }
interface ApifyPlatformApi { platform: string; runs: number; items_scraped: number; cost_usd: number }
interface ApifyDailyApi { day: string; total: number; by_platform: Record<string, number> }
interface ApifyRollupApi {
  kpis: ApifyKpisApi
  by_source: ApifySourceApi[]
  by_account: ApifyAccountApi[]
  by_platform: ApifyPlatformApi[]
  daily: ApifyDailyApi[]
  platforms: string[]
}
interface ApifyRunApi {
  id: number
  created_at: string
  platform: string
  account: string
  actor_id: string
  apify_run_id: string | null
  status: string
  items_scraped: number
  items_kept: number
  cost_usd: number | null
  charged_events: Record<string, number> | null
  source_id: number | null
  source_name: string | null
  ingestion_run_id: string | null
}

function toApifyRollup(r: ApifyRollupApi): ApifyRollup {
  return {
    kpis: {
      costUsd: r.kpis.cost_usd,
      runs: r.kpis.runs,
      itemsScraped: r.kpis.items_scraped,
      itemsKept: r.kpis.items_kept,
      errors: r.kpis.errors,
      accounts: r.kpis.accounts,
      prevCostUsd: r.kpis.prev_cost_usd,
      prevRuns: r.kpis.prev_runs,
    },
    bySource: r.by_source.map((s) => ({
      sourceId: s.source_id,
      sourceName: s.source_name,
      runs: s.runs,
      itemsScraped: s.items_scraped,
      costUsd: s.cost_usd,
    })),
    byAccount: r.by_account.map((a) => ({
      platform: a.platform,
      account: a.account,
      runs: a.runs,
      itemsScraped: a.items_scraped,
      costUsd: a.cost_usd,
    })),
    byPlatform: r.by_platform.map((p) => ({
      platform: p.platform,
      runs: p.runs,
      itemsScraped: p.items_scraped,
      costUsd: p.cost_usd,
    })),
    daily: r.daily.map((d) => ({ day: d.day, total: d.total, byPlatform: d.by_platform })),
    platforms: r.platforms,
  }
}

function toApifyRun(r: ApifyRunApi): ApifyRunRow {
  return {
    id: r.id,
    createdAt: r.created_at,
    platform: r.platform,
    account: r.account,
    actorId: r.actor_id,
    apifyRunId: r.apify_run_id,
    status: r.status,
    itemsScraped: r.items_scraped,
    itemsKept: r.items_kept,
    costUsd: r.cost_usd,
    chargedEvents: r.charged_events,
    sourceId: r.source_id,
    sourceName: r.source_name,
    ingestionRunId: r.ingestion_run_id,
  }
}

/** Rollup del gasto Apify del rango (KPIs + por fuente/cuenta/plataforma + serie diaria). */
export async function fetchApifyRollup(w: MetricsWindow = {}): Promise<ApifyRollup> {
  const qs = new URLSearchParams()
  windowParams(qs, w)
  return toApifyRollup(await apiGet<ApifyRollupApi>(`/metrics/apify/rollup?${qs.toString()}`))
}

/** Corridas de actor crudas (auditoría), más recientes primero. */
export async function fetchApifyRuns(
  w: MetricsWindow = {},
  opts: { limit?: number; offset?: number } = {},
): Promise<{ items: ApifyRunRow[]; total: number }> {
  const qs = new URLSearchParams()
  windowParams(qs, w)
  qs.set("limit", String(opts.limit ?? 50))
  qs.set("offset", String(opts.offset ?? 0))
  const r = await apiGet<{ items: ApifyRunApi[]; total: number }>(
    `/metrics/apify/runs?${qs.toString()}`,
  )
  return { items: r.items.map(toApifyRun), total: r.total }
}

export type FilterMode = "include" | "exclude"

export interface LlmCallsQuery extends MetricsWindow {
  status?: string[]
  statusMode?: FilterMode
  module?: string[]
  moduleMode?: FilterMode
  model?: string[]
  modelMode?: FilterMode
  /** Filtra por source_name (incluye los pseudo "(calendar)"/"(sin source)"). */
  source?: string[]
  sourceMode?: FilterMode
  q?: string
  sort?: "created_at" | "cost_usd" | "latency_ms"
  dir?: "asc" | "desc"
  limit?: number
  offset?: number
}

export function multiParam(
  qs: URLSearchParams,
  name: string,
  values: string[] | undefined,
  mode: FilterMode | undefined,
): void {
  if (!values || values.length === 0) return
  for (const v of values) qs.append(name, v)
  if (mode === "exclude") qs.set(`${name}_mode`, "exclude")
}

/** Filas crudas de auditoría con filtros incluir/excluir, orden y paginación offset. */
export async function fetchLlmCalls(query: LlmCallsQuery): Promise<{ items: LlmCallRow[]; total: number }> {
  const qs = new URLSearchParams()
  windowParams(qs, query)
  multiParam(qs, "status", query.status, query.statusMode)
  multiParam(qs, "module", query.module, query.moduleMode)
  multiParam(qs, "model", query.model, query.modelMode)
  multiParam(qs, "source", query.source, query.sourceMode)
  if (query.q) qs.set("q", query.q)
  if (query.sort) qs.set("sort", query.sort)
  if (query.dir) qs.set("dir", query.dir)
  qs.set("limit", String(query.limit ?? 50))
  qs.set("offset", String(query.offset ?? 0))
  const r = await apiGet<{ items: CallApi[]; total: number }>(`/metrics/llm/calls?${qs.toString()}`)
  return { items: r.items.map(toRow), total: r.total }
}
